src/Arcus.Messaging.Pumps.Abstractions/MessagePump.cs

```diff
@@ -83,62 +82,80 @@ protected virtual Task HandleReceiveExceptionAsync(Exception receiveException)
+        ///     Handle a new message that was received
         /// </summary>
         /// <param name="message">Message that was received</param>
         /// <param name="messageContext">Context providing more information concerning the processing</param>
         /// <param name="correlationInfo">
-        ///     Information concerning correlation of telemetry & processes by using a variety of unique
+        ///     Information concerning correlation of telemetry and processes by using a variety of unique
         ///     identifiers
         /// </param>
         /// <param name="cancellationToken">Cancellation token</param>
-        protected abstract Task ProcessMessageAsync(TMessage message, TMessageContext messageContext,
-            MessageCorrelationInfo correlationInfo, CancellationToken cancellationToken);
-
-        /// <summary>
-        ///     Determines the encoding used for a given message
-        /// </summary>
-        /// <remarks>If no encoding was specified, UTF-8 will be used by default</remarks>
-        /// <param name="messageContext">Context concerning the message</param>
-        /// <returns>Encoding that was used for the message body</returns>
-        protected virtual Encoding DetermineMessageEncoding(MessageContext messageContext)
+        protected async Task ProcessMessageAsync<TMessageHandler>(
+            TMessage message,
+            TMessageContext messageContext,
+            MessageCorrelationInfo correlationInfo,
+            CancellationToken cancellationToken)
+            where TMessageHandler : IMessageHandler<TMessage, TMessageContext>
         {
-            if (messageContext.Properties.TryGetValue(PropertyNames.Encoding, out object annotatedEncoding))
+            IEnumerable<TMessageHandler> messageHandlers = ServiceProvider.GetServices<TMessageHandler>();
+            if (messageHandlers is null || !messageHandlers.Any())
+            {
+                throw new InvalidOperationException(
+                    $"Message pump cannot correctly process the '{typeof(TMessage).Name}' in the '{typeof(TMessageContext)}' "
+                    + $"because no '{nameof(IMessageHandler<TMessage, TMessageContext>)}' was registered in the dependency injection container. "
+                    + $"Make sure you call the correct '.With...' extension on the {nameof(IServiceCollection)} during the registration of the message pump to register a message handler");
+            }
+
+            foreach (TMessageHandler messageHandler in messageHandlers)
             {
-                try
+                if (messageHandler is null)
                 {
-                    return Encoding.GetEncoding(annotatedEncoding.ToString());
+                    continue;
                 }
-                catch (Exception ex)
+
+                bool isProcessed = await TryProcessMessageAsync(messageHandler, message, messageContext, correlationInfo, cancellationToken);
```

Comment:
This is tricky, as discussed we should:
1. Try to deserialize the message according to the expected contract of the handler
2. If it matches, only the first match should process it
3. Done

Otherwise we're at risk of running a lot of duplicate processing, which failed, but still can cause inconsistent state